### PR TITLE
Reddit Post Position Config & Download Progress Bar

### PR DESCRIPTION
**Pull request by William9923**
### Description
Added:
- Dynamic image clip position for each available background
- Progress bar when downloading youtube video
- Few additional background video (GTA, Rocket League, Assasin Creed) as example

### Why? (Motivation)
Although the bot has produced a lot of good videos with minecraft background, I noticed that on some other background, the image of reddit post was blocking the background (check below image). So, it inspire me to make different configuration for each background. For now, it would be the reddit post image position on the background

<table>
  <tr>
    <td valign="top">
        before
    </td>
    <td valign="top">
        after
    </td>
  </tr>
  <tr>
    <td><p align="center">Before</p></td>
    <td><p align="center">After</p></td>
  </tr>
</table>

Also, when trying out different background, I noticed that we need to download all background video simultaneously. Because the time it takes to download all video is quite long, I propose that the bot only need to download for needed video & add the progress bar to track the download progress.

<table>
  <tr>
    <td valign="top">
        During
    </td>
    <td valign="top">
        End
    </td>
  </tr>
  <tr>
    <td><p align="center">Downloading...</p></td>
    <td><p align="center">Done</p></td>
  </tr>
</table>

### How (?)
For changing the position of reddit post image for each background, I use the available `set_position` method in `moviepy`. After refactoring the code for the background video part, we can further make separate configuration (not only the position of reddit post image), such as the blur effect (opacity) etc. 

For the progress bar, I simply add tqdm progress bar and use `on_progress_callback` that was available from pytube API when downloading the video.

### Test Guide
To make testing the new feature easier, you could follow this step by step guides below (i don't include other functionality test, but it should be the same as other functionality):
1. Testing out different screenshot placement in background.
 - First, might need to read [this](https://zulko.github.io/moviepy/ref/VideoClip/VideoClip.html#moviepy.video.VideoClip.VideoClip.set_position). It contains reference on how moviepy put the image in the background clips
 - Second, try out different background to see the diff. (example: minecraft use `center` , while other like GTA Stunt race move the screenshot slightly higher)
 - Third, feel free to modify the position for each background.
 

2. Testing out the new progress bar during Youtube video download
- First, run the program with background that you never downloaded (or just remove the background vids)
- Second, wait till the download done. You will see the progress bar during download...

3. Testing out new random background (default, if not filled or unsupported background)
- Don't fill the .env variable `BackgroundChoice`. It will use random video (from supported vids) as background

4. Adding new background (this is not new feature, but this PR slightly alter how to add new video as background)

### CODE REVIEW

**Comment by IslaFTW** [comment]
> ### Description
> Added:
> 
> * Dynamic image clip position for each available background
> * Progress bar when downloading youtube video
> * Few additional background video (GTA, Rocket League, Assasin Creed) as example
> 
> ### Why? (Motivation)
> Although the bot has produced a lot of good videos with minecraft background, I noticed that on some other background, the image of reddit post was blocking the background (check below image). So, it inspire me to make different configuration for each background. For now, it would be the reddit post image position on the background
> 
> before	after
> Before
> 
> After
> 
> Also, when trying out different background, I noticed that we need to download all background video simultaneously. Because the time it takes to download all video is quite long, I propose that the bot only need to download for needed video & add the progress bar to track the download progress.
> 
> During	End
> Downloading...
> 
> Done
> 
> ### How (?)
> For changing the position of reddit post image for each background, I use the available `set_position` method in `moviepy`. After refactoring the code for the background video part, we can further make separate configuration (not only the position of reddit post image), such as the blur effect (opacity) etc.
> 
> For the progress bar, I simply add tqdm progress bar and use `on_progress_callback` that was available from pytube API when downloading the video.

Hello,

I'm not a developer, but I've been using this project for a week, and I try to help at least with ideas and tests, among other basic things.

I was reading your PR and just this week I was thinking about this improvement.

I thought maybe you could see the following:

Find out what the official TikTok "safe zones" are, and it would be great to be able to move the screenshot also horizontally, since currently, as the TikTok interface is published, it covers part of the Screenshot, I leave you an example:
289022985_722387272427917_353594956416899694_n

On the official website of tiktok, there are the rules of the safezones so that your content is not covered by its tiktok interface:
https://ads.tiktok.com/help/article?aid=10002742

basically there is 120px on the right side that contains buttons from the tiktok interface, it would be great if the screenshots respect that 120px on the right side

thumbnail_e9dd6157a46442e69b5ea15f28d24ab5_jpg

I hope I didn't bother you with this, but your PR seemed the perfect thing to solve this detail too

**Comment by lamasterso** [comment]
@IslaFTW  Bro Do you know how to make the text size a little bit larger ?

**Comment by IslaFTW** [comment]
> @IslaFTW Bro Do you know how to make the text size a little bit larger ?

I think it is not possible, but it would be great to be able to control the size of the font and also the type of Font, for example the "Ubuntu" font in size 12, it is very pleasant to read, and it is very beautiful

**Comment by lamasterso** [comment]
I tried to resize the screenshot and did everything but there is no variable to control the Height Of the screenshot, Therefore I was only able to manage changing the width, But I think making the height of the screenshot more will stretch the font somehow. 

Now the font is very wide but in very small letters to read especially the comments.

**Comment by William9923** [comment]
@lamasterso One of the solution that I can think of is resizing the image to ensure that the image respect the Tiktok interface (120px). I will try to see what I can do on this. Thank you for your suggestion 👍 .

I probably need time to research if the size (120px) is the same for all device or depends on the device.

**Comment by JasonLovesDoggo** [comment]
This is great and all but you are by far over complicating things, you don't need separate files and classes for every single background and you don't need a background swapper class

Also @IslaFTW suggestion was great

I will merge this into a separate branch and improve this there in 3 days

**Comment by JasonLovesDoggo** [comment]
> Hi @JasonLovesDoggo , I have adjusted this PR with the recent code change and refactor `background.py` for downloading background video from Youtube parts to simplify it. Would you be kindly enough to review the changes again. Thank you 🙏 

I will look at it when I get home around 
6:00 p.m. EST

**Comment by CJthisis** [comment]
Hey there, I tried to run the code in combination with the latest develop build, but it doesn't work.

Specifically, when trying to download the background videos, in the line 19 of background.py, there is is the name 'Progress', which does't seem to appear anywhere else in the file? 
grafik

**Comment by JasonLovesDoggo** [comment]
> Hey there, I tried to run the code in combination with the latest develop build, but it doesn't work.
> 
> Specifically, when trying to download the background videos, in the line 19 of background.py, there is is the name 'Progress', which does't seem to appear anywhere else in the file? 
> I don't know how to fix this, maybe you know what to do?
> 
> grafik
> 

I will take a look at that

**Comment by William9923** [comment]
> Hey there, I tried to run the code in combination with the latest develop build, but it doesn't work. 

Specifically, when trying to download the background videos, in the line 19 of background.py, there is is the name 'Progress', which does't seem to appear anywhere else in the file?
I don't know how to fix this, maybe you know what to do?

I will help to look into it too. Maybe it related on the new progress bar that were included in this PR.

**Comment by William9923** [comment]
> Hey there, I tried to run the code in combination with the latest develop build, but it doesn't work.
> 
> Specifically, when trying to download the background videos, in the line 19 of background.py, there is is the name 'Progress', which does't seem to appear anywhere else in the file? I don't know how to fix this, maybe you know what to do?
> 
> grafik

**Comment by William9923** [comment]
Hi @CJthisis , I have updated my code to the newest develop branch (and tested it!). The problem from your screenshot should not happen anymore, and would you kindly enough to try it again? Thank youu 🙏 .

**Comment by JasonLovesDoggo** [comment]
This is great but I just have one last feature request from this Have it so that if the background choice option is empty It will pick a random background

**Comment by JasonLovesDoggo** [comment]
This is great but I just have one last feature request from this Have it so that if the background choice option is empty It will pick a random background q

**Comment by CJthisis** [comment]
> This is great but I just have one last feature request from this Have it so that if the background choice option is empty It will pick a random background

This would be great!

**Comment by CJthisis** [comment]
> Hi @CJthisis and @JasonLovesDoggo  , I have updated my code to the newest develop branch (and tested it!). The problem from your screenshot should not happen anymore, and would you kindly enough to try it again? Thank youu 🙏 .
> 
> 
> 
> 

I will do so tomorrow when I get home. Do you have updated your fork too? If so, I would use that version

**Comment by William9923** [comment]
> 



> This is great but I just have one last feature request from this Have it so that if the background choice option is empty It will pick a random background

Will do, thanks for the suggestion

**Comment by William9923** [comment]
Hi, I've implemented the default (if user not pick any background) background == random background (that the bot support). I will write a guide to test it out so you guys have an easier time trying the new feature. Thanks 🙏

**Comment by CordlessCoder** [comment]
Will test and review when I get home

**Comment by JasonLovesDoggo** [comment]
@CordlessCoder If this works for you. Merge it

**Comment by William9923** [comment]
@CordlessCoder hi, due to new update in develop branch, I need to adjust a few line of codes. Please update your branch (or just delete and pull it again haha) when reviewing. Thankss 🙏

**Comment by William9923** [comment]
bump @JasonLovesDoggo , @CordlessCoder any update on this PR 😄 ? A little bit tired of updating it to the newest develop branch every day haha.

**Comment by JasonLovesDoggo** [comment]
The thing is we are about to change from the ENV file to a different system so it would be incompatible

**Comment by CallumIO** [comment]
> bump @JasonLovesDoggo , @CordlessCoder any update on this PR smile ? A little bit tired of updating it to the newest develop branch every day haha.

Once the new config system is in, I'll fix the merge conflicts and test

**Comment by William9923** [comment]
> The thing is we are about to change from the ENV file to a different system so it would be incompatible

I see, thx for the heads up.

**Comment by William9923** [comment]
> > bump @JasonLovesDoggo , @CordlessCoder any update on this PR smile ? A little bit tired of updating it to the newest develop branch every day haha.
> 
> Once the new config system is in, I'll fix the merge conflicts and test

sure", thx btw. I'm always available incase any issue comes up when merging 🙏 .
